### PR TITLE
PMM-9235 notify if source branches are outdated

### PR DESCRIPTION
**Pull request by yurkovychv**
https://jira.percona.com/browse/PMM-9235

### CODE REVIEW

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Here is the list:
 https://github.com/percona/pmm-ui-tests/pull/111

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Here is the list:
 https://github.com/percona/pmm-ui-tests/pull/40

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Here is the list:
 https://github.com/percona/pmm-ui-tests/pull/40

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Please update them and restart the job
 https://github.com/percona/pmm-ui-tests/pull/40
 Corresponding artifacts will be produced but without executing tests

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Please update them and restart the job
 https://github.com/percona/pmm-ui-tests/pull/40

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Please update them and restart the job
 https://github.com/percona/pmm-ui-tests/pull/40

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Please update them and restart the job
 https://github.com/percona/pmm-ui-tests/pull/40

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Please update them and restart the job
 https://github.com/percona/pmm-ui-tests/pull/137

**Comment by JNKPercona** [comment]
Looks like there are outdated source branches.
 Please update them and restart the job
 https://github.com/percona/pmm-ui-tests/pull/108

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-2167-6f42c86
client docker - perconalab/pmm-client-fb:PR-2167-6f42c86
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2167-6f42c86.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-2167-6f42c86&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2167-6f42c86.tar.gz

**Comment by JNKPercona** [comment]
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/8654/

**Comment by JNKPercona** [comment]
pmm2-client testsuite has failed, Please check: BATS: https://pmm.cd.percona.com/job/pmm2-testsuite/8102/

File: ci.py [hunk 1]
```diff
@@ -167,6 +168,41 @@ def repo_cloned():
                 f.write(f'export {dep_name_underscore}_commit={commit_id}')
                 f.write(f'export {dep_name_underscore}_branch={dep["branch"]}\n')
 
+    def check_deps(self):
+        outdated_branches_message = 'Looks like there are outdated source branches.\n Please update them and restart ' \
+                                    'the job'
+        outdated_branches = []
+        submodules_url = '/'.join(PR_URL.split('/')[3:-2])
+        pull_number = PR_URL.split('/')[-1:][0]
```
Comment (nikita-b) [inline ci.py:177]: Please add example for CHANGE_URL here :)

File: ci.py [hunk 2]
```diff
@@ -167,6 +168,41 @@ def repo_cloned():
...
+        
+        if GITHUB_TOKEN == '':
+            print('there is no GITHUB_TOKEN')
```
Comment (nikita-b) [inline ci.py]:
```suggestion
            logging.warning('there is no GITHUB_TOKEN')
```

File: ci.py [hunk 3]
```diff
@@ -167,6 +168,41 @@ def repo_cloned():
...
+
+        github_api = Github(GITHUB_TOKEN)
+
+        for dep in self.config_override['deps']:
+            target_url = dep['url']
+            repo_path = '/'.join(target_url.split('/')[-2:])
+            target_branch = dep['branch']
+            r = github_api.get_repo(repo_path)
```
Comment (nikita-b) [inline ci.py]: It's better to use the full name for variables. i.e repo

File: ci.py [hunk 4]
```diff
@@ -167,6 +168,41 @@ def repo_cloned():
...
+            if not pulls_list.totalCount:
+                continue
+
+            pull = r.get_pull(pulls_list[0].number)
+            if pull.mergeable_state in ['behind', 'dirty']:
+                outdated_branches.append(pull.html_url)
+
+        if outdated_branches:
+            for branch_url in outdated_branches:
+                outdated_branches_message = f'{outdated_branches_message}\n {branch_url}'
```
Comment (nikita-b) [inline ci.py]:
```suggestion
                outdated_branches_message += f'\n {branch_url}'
```

File: ci.py [hunk 5]
```diff
@@ -167,6 +168,41 @@ def repo_cloned():
...
+                
+            r = github_api.get_repo(submodules_url)
```
Comment (nikita-b) [inline ci.py]: r to repo

**Review by nikita-b — CHANGES_REQUESTED** [review]

File: ci.py (hunk 1 above)
Comment (yurkovychv) [reply ci.py:177]: added above, thanks

File: ci.py (hunk 3 above)
Comment (yurkovychv) [reply ci.py]: done

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-2167-521c258
client docker - perconalab/pmm-client-fb:PR-2167-521c258
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2167-521c258.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-2167-521c258&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2167-521c258.tar.gz

**Comment by JNKPercona** [comment]
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/8671/

**Comment by JNKPercona** [comment]
server docker - perconalab/pmm-server-fb:PR-2167-d61d5b3
client docker - perconalab/pmm-client-fb:PR-2167-d61d5b3
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2167-d61d5b3.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-2167-d61d5b3&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2167-d61d5b3.tar.gz

**Comment by JNKPercona** [comment]
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/11007/

**Review by nikita-b — APPROVED** [review]

**Review by puneet0191 — APPROVED** [review]
thanks a lot Vasya

**Pull request merged**